common: plugin framework

pbalcer opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/2179)
<!-- Reviewable:end -->

krzycz commented:
Reviewed 14 of 14 files at r1.
Review status: 7 of 14 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/common/plugin.c, line 46 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6JnTZOOggVBMP12v1:-Kr6JnTZOOggVBMP12v2:b-cx7z0n) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/common/plugin.c#L46)):*
> ```C
> 
> struct plugin {
> 	const char *base_name;
> ```

Any description of those fields? I.e. what's the meaning/difference between *base_name* and *name*?

---

*[src/common/plugin.c, line 91 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6NvcbB3Fjh8oTMT9t:-Kr6NvccDUKXJrhjaIXH:bppn87) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/common/plugin.c#L91)):*
> ```C
> 	}
> 
> 	p->loaded = 0;
> ```

is it ever set to 1?

---

*[src/common/plugin.c, line 147 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6KaqSzAaGOY5b-wRH:-Kr6KaqTxay5mmX55KaD:b-7oulqv) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/common/plugin.c#L147)):*
> ```C
>  * plugin_init -- initializes plugin module
>  */
> void
> ```

int ?

---

*[src/common/plugin.c, line 163 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6KWHsFE6H966dA_mZ:-Kr6KWHsFE6H966dA_m_:b-qwek6x) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/common/plugin.c#L163)):*
> ```C
> 	if (plugin_path == NULL) {
> 		closedir(pdir);
> 		return;
> ```

Why not ERR() or even FATAL()?

---

*[src/common/plugin.c, line 205 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6OTyfmNbwcSFuAADH:-Kr6OTyfmNbwcSFuAADI:b-6oleat) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/common/plugin.c#L205)):*
> ```C
> }
> 
> void
> ```

missing brief

---

*[src/test/util_plugin/plugin0.c, line 71 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6PNhFQ28oQkTvtyEb:-Kr6PNhFQ28oQkTvtyEc:b260uco) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/test/util_plugin/plugin0.c#L71)):*
> ```C
> int pmem_plugin_funcs(void **funcs)
> {
> 	*funcs = &plugin0_funcs;
> ```

Why not return it by value?

---

*[src/test/util_plugin/plugin0.c, line 73 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6P7kmUxDCr_-cWdzp:-Kr6P7kmUxDCr_-cWdzq:b-4zktzv) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/test/util_plugin/plugin0.c#L73)):*
> ```C
> 	*funcs = &plugin0_funcs;
> 
> 	return 0;
> ```

Can it return non-zero in any plugin?

---

*[src/test/util_plugin/util_plugin.c, line 75 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6Q45iQul8ujDjsTu4:-Kr6Q45iQul8ujDjsTu5:byastqh) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/test/util_plugin/util_plugin.c#L75)):*
> ```C
> 	START(argc, argv, "util_plugin");
> 
> 	plugin_init(".");
> ```

As I understand the design, to use plugin libraries you need to:
1) scan the specified directory for all the plugins and register them
2) explicitly load the required plugin(s) by *base_name* (which loads all the plugins from given "group")

Also, I can call *plugin_init* multiple times, to load plugins from multiple locations, right?

Q1a: What happens if there is more than one plugin with the same base_name/name?
Q1b: What if they have the same/different version?

Q2a: Wouldn't it be easier to pass a list of locations (directories) to *plugin_init* to register all plugins in a single call?
Q2b: The alternate approach could be to to register plugin directories only in *plugin_init*, and scan those directories at *plugin_load*.

Q3: I can imagine someone may want to load two plugins of the same "class", but different version/implementation at the same time. Would it be possible?
For example, if (at some point) we have a "pmem_allocator" plugin that implements PMEM allocator APIs, one may have two different implementations loaded at the same time, used in one program, but in two different pools. I'm not sure if name/version is enough to identify the correct version of the plugin. Perhaps a UUID would be better?
For me, "name" identifies a plugin "class" (i.e. "pmem_allocator"); version denotes the list of the plugin APIs (exported funcs), which must be compatible with the version of pmemobj library; and we need some identifier for the "instance"/"implementation variant" of the plugin.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2179)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2179)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[src/common/plugin.c, line 70 at r2](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr7ApmVhYchyiNeK0Wu:-Kr7ApmVhYchyiNeK0Wv:b7hqv1w) ([raw file](https://github.com/pmem/nvml/blob/7324b4aead20fbc33a86435d93d0863814329e0d/src/common/plugin.c#L70)):*
> ```C
> static SLIST_HEAD(, plugin) plugins;
> 
> #define PMEM_PLUGIN_LOAD_SYMBOL(plugin, symbol)\
> ```

Does this need to be a macro?

---

*[src/common/plugin.c, line 107 at r2](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr7EeyjbQZL5NEP9FC1:-Kr7EeyjbQZL5NEP9FC2:b43pfb4) ([raw file](https://github.com/pmem/nvml/blob/7324b4aead20fbc33a86435d93d0863814329e0d/src/common/plugin.c#L107)):*
> ```C
> 	PMEM_PLUGIN_LOAD_SYMBOL(p, pmem_plugin_unload);
> 
> 	p->base_name = p->p_ops.pmem_plugin_base_name();
> ```

Is it really useful to allow the plugin to pick a name runtime?

Wouldn't it suffice, to just:
```
p->base_name = util_dlsym(p->handle, "pmem_plugin_base_name");
```
Same question about plugin_name, plugin_version.

---

*[src/common/plugin.c, line 174 at r2](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr7D8Y-bUM0vwC8h_jB:-Kr7D8Y-bUM0vwC8h_jC:b-rnel5b) ([raw file](https://github.com/pmem/nvml/blob/7324b4aead20fbc33a86435d93d0863814329e0d/src/common/plugin.c#L174)):*
> ```C
> 			continue;
> 
> 		strcpy(plugin_path + dirlen + 1, entry->d_name);
> ```

This could look nicer using fts(3)

No need to Malloc and concatenate strings.

Something like:
```
while ((entry = fts_read(f)) != NULL) {
   if (entry->fts_namelen > 3 && strcmp(entry->fts_name + entry->fts_namelen - 3, ".so") == 0)
       plugin_new_entry(entry->fts_path);
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2179)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[src/common/plugin.c, line 91 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6NvcbB3Fjh8oTMT9t:-Kr9y4XtFbrIasTGbJFZ:b-r1ij3j) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/common/plugin.c#L91)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

is it ever set to 1?
</blockquote></details>

Good catch ;)

---

*[src/common/plugin.c, line 163 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6KWHsFE6H966dA_mZ:-Kr9z0St2RSurZMpAFKZ:b1cyzi0) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/common/plugin.c#L163)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Why not ERR() or even FATAL()?
</blockquote></details>

Error handling should always be handled in the upper most layers.

---

*[src/common/plugin.c, line 205 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6OTyfmNbwcSFuAADH:-Kr9zDfI7CRv4sgizNe5:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/common/plugin.c#L205)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

missing brief
</blockquote></details>

Done.

---

*[src/common/plugin.c, line 70 at r2](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr7ApmVhYchyiNeK0Wu:-Kr9xwLWGjKN7ewLMR3O:bbfp63m) ([raw file](https://github.com/pmem/nvml/blob/7324b4aead20fbc33a86435d93d0863814329e0d/src/common/plugin.c#L70)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

Does this need to be a macro?
</blockquote></details>

How would you do it differently?
I use the symbol both as a string and as a field in a struct. Can't do that in a function.

---

*[src/common/plugin.c, line 107 at r2](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr7EeyjbQZL5NEP9FC1:-Kr9yDl-bNujXMHdOHMh:bj16vtr) ([raw file](https://github.com/pmem/nvml/blob/7324b4aead20fbc33a86435d93d0863814329e0d/src/common/plugin.c#L107)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

Is it really useful to allow the plugin to pick a name runtime?

Wouldn't it suffice, to just:
```
p->base_name = util_dlsym(p->handle, "pmem_plugin_base_name");
```
Same question about plugin_name, plugin_version.
</blockquote></details>

I'm not sure if I understand the question.
Do you mean that the code should omit calling these functions here and instead call them whenever the field is actually read?
The reason I did it this way is because I want to avoid adding potentially broken plugins (i.e., those that return NULL as name) to the list.

---

*[src/common/plugin.c, line 174 at r2](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr7D8Y-bUM0vwC8h_jB:-Kr9zGbih7038QEGQiFn:bp76sp0) ([raw file](https://github.com/pmem/nvml/blob/7324b4aead20fbc33a86435d93d0863814329e0d/src/common/plugin.c#L174)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

This could look nicer using fts(3)

No need to Malloc and concatenate strings.

Something like:
```
while ((entry = fts_read(f)) != NULL) {
   if (entry->fts_namelen > 3 && strcmp(entry->fts_name + entry->fts_namelen - 3, ".so") == 0)
       plugin_new_entry(entry->fts_path);
}
```
</blockquote></details>

But that's also less portable.
I can change it, but it would mean I have to implement something different for Windows.

---

*[src/test/util_plugin/plugin0.c, line 73 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6P7kmUxDCr_-cWdzp:-KrA-kGmyxw9YiEGdzDZ:bo58ss1) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/test/util_plugin/plugin0.c#L73)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Can it return non-zero in any plugin?
</blockquote></details>

Depends, but I guess it can't.

---

*[src/test/util_plugin/util_plugin.c, line 75 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6Q45iQul8ujDjsTu4:-Kr9zNbYab1hQQhPM3dX:b7ldxqx) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/test/util_plugin/util_plugin.c#L75)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

As I understand the design, to use plugin libraries you need to:
1) scan the specified directory for all the plugins and register them
2) explicitly load the required plugin(s) by *base_name* (which loads all the plugins from given "group")

Also, I can call *plugin_init* multiple times, to load plugins from multiple locations, right?

Q1a: What happens if there is more than one plugin with the same base_name/name?
Q1b: What if they have the same/different version?

Q2a: Wouldn't it be easier to pass a list of locations (directories) to *plugin_init* to register all plugins in a single call?
Q2b: The alternate approach could be to to register plugin directories only in *plugin_init*, and scan those directories at *plugin_load*.

Q3: I can imagine someone may want to load two plugins of the same "class", but different version/implementation at the same time. Would it be possible?
For example, if (at some point) we have a "pmem_allocator" plugin that implements PMEM allocator APIs, one may have two different implementations loaded at the same time, used in one program, but in two different pools. I'm not sure if name/version is enough to identify the correct version of the plugin. Perhaps a UUID would be better?
For me, "name" identifies a plugin "class" (i.e. "pmem_allocator"); version denotes the list of the plugin APIs (exported funcs), which must be compatible with the version of pmemobj library; and we need some identifier for the "instance"/"implementation variant" of the plugin.

</blockquote></details>

Q1a: if they are of different versions, you can still distinguish between them. If they are not, you get two plugins loaded. I think that's fine. The 'name' is information only, we will never expect certain plugins to be loaded.
Q1b: Depends on the business logic in the upper layer.

Q2a: It wouldn't be easier. You can just as easily call this function in a loop. I expect the 90% use case to be a single call with a single directory at the library load. The other use cases would be loading plugins through CTL. Which would be a separate call.
Q2b: Error handling would be more difficult.

Q3: The 'base_name' is the 'class'. The 'name' is unique identifier. In your example, 'base_name' would be 'pmem_allocator' and the 'name' would be 'my_awesome_alloc'. But again, notice that we never compare the plugin name to anything - it's just info.

As for the UUID, the plugin author would still have to pick a static number anyway. Might as well just pick a static human-readable string.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2179)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: 4 of 13 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/common/plugin.c, line 107 at r2](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr7EeyjbQZL5NEP9FC1:-KrAGc61AdV9hghDn91H:b-bc32qq) ([raw file](https://github.com/pmem/nvml/blob/7324b4aead20fbc33a86435d93d0863814329e0d/src/common/plugin.c#L107)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

I'm not sure if I understand the question.
Do you mean that the code should omit calling these functions here and instead call them whenever the field is actually read?
The reason I did it this way is because I want to avoid adding potentially broken plugins (i.e., those that return NULL as name) to the list.
</blockquote></details>

I didn't talk about checking for NULL, let me extend that example:
```
if ((p->base_name = util_dlsym(p->handle, "pmem_plugin_base_name")) == NULL)
	LOG(3, "%s: unable to find base_name symbol", path);
```
The question was about why the symbols need to refer to functions, why not just to C strings?

---

*[src/common/plugin.c, line 174 at r2](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr7D8Y-bUM0vwC8h_jB:-KrAHeBzLGg_SbulnT2D:b-ej5j6b) ([raw file](https://github.com/pmem/nvml/blob/7324b4aead20fbc33a86435d93d0863814329e0d/src/common/plugin.c#L174)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

But that's also less portable.
I can change it, but it would mean I have to implement something different for Windows.
</blockquote></details>

Ah, I didn't  know it was meant to work on windows too. Is there opendir, and closedir on windows?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2179)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Review status: 4 of 13 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/common/plugin.c, line 46 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6JnTZOOggVBMP12v1:-KrAC3JiYl9qiU4gQgpJ:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/common/plugin.c#L46)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Any description of those fields? I.e. what's the meaning/difference between *base_name* and *name*?
</blockquote></details>

Done.

---

*[src/common/plugin.c, line 147 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6KaqSzAaGOY5b-wRH:-KrA-rnBeGUgdyOQk79r:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/common/plugin.c#L147)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

int ?
</blockquote></details>

Done.

---

*[src/common/plugin.c, line 107 at r2](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr7EeyjbQZL5NEP9FC1:-KrAIqyvfDAtr_HCjPby:b3dns9y) ([raw file](https://github.com/pmem/nvml/blob/7324b4aead20fbc33a86435d93d0863814329e0d/src/common/plugin.c#L107)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

I didn't talk about checking for NULL, let me extend that example:
```
if ((p->base_name = util_dlsym(p->handle, "pmem_plugin_base_name")) == NULL)
	LOG(3, "%s: unable to find base_name symbol", path);
```
The question was about why the symbols need to refer to functions, why not just to C strings?
</blockquote></details>

Gotcha.
That's how I've usually done it and seen it done.

I've changed it now anyway so the issue is no longer relevant ;)

---

*[src/common/plugin.c, line 174 at r2](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr7D8Y-bUM0vwC8h_jB:-KrAJbpIcvCLmGoQmZcQ:bkdo8hu) ([raw file](https://github.com/pmem/nvml/blob/7324b4aead20fbc33a86435d93d0863814329e0d/src/common/plugin.c#L174)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

Ah, I didn't  know it was meant to work on windows too. Is there opendir, and closedir on windows?
</blockquote></details>

I hope so :D

If there isn't, I'm going to implement it the way you suggested.

---

*[src/test/util_plugin/plugin0.c, line 71 at r1](https://reviewable.io:443/reviews/pmem/nvml/2179#-Kr6PNhFQ28oQkTvtyEb:-KrAAwqj9FvH5cRcQJV-:b-cu170v) ([raw file](https://github.com/pmem/nvml/blob/bb45ad08668678473abc9f50bbbb685b9c5e4dfb/src/test/util_plugin/plugin0.c#L71)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Why not return it by value?
</blockquote></details>

Changed this a bit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2179)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 9 of 10 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2179#-:-KrARsd8l9KoyiwV9UjJ:bnfp4nl)*
<!-- Sent from Reviewable.io -->
